uncertainties/unumpy/core.py

```diff
@@ -30,10 +30,87 @@
     # Utilities:
     "nominal_values",
     "std_devs",
+    "average",
     # Classes:
     "matrix",
 ]
 
+
+def _flatten_array_by_axes(arr, axes):
+    """
+    Return arr, reshaped such that the axes listed in parameter ``axes`` are
+    flattened, and become the last axis of the resulting array. A utility used
+    for `:func:uncertainties.unumpy.average`.
+    """
+    arr = np.asanyarray(arr)
+    if axes is None:
+        return arr.flatten()
+    # The following sanity checks on axes can be replaced by
+    # np.lib.array_utils.normalize_axis_tuple once we will require a minimum
+    # version of numpy 2.x.
+    if type(axes) not in (tuple, list):
+        try:
+            axes = [operator.index(axes)]
+        except TypeError:
+            pass
+    axes = tuple(axes)
+    if len(set(axes)) != len(axes):
+        raise ValueError("repeated axis found in axes argument")
+    # This one is not checked by np.lib.array_utils.normalize_axis_tuple
+    if max(axes) >= arr.ndim:
+        raise ValueError(
+            f"Cannot average over an inexistent axis {max(axes)} >= arr.ndim = "
+            f"{arr.ndim}"
+        )
+    new_shape = []
+    # To hold the product of the dimensions to flatten
+    flatten_size = 1
+    for i in range(len(arr.shape)):
+        if i in axes:
+            flatten_size *= arr.shape[i]  # Multiply dimensions to flatten
+        else:
+            new_shape.append(arr.shape[i])  # Keep the dimension
+    # This way the shapes to average over are flattend, in the end.
+    new_shape.append(flatten_size)
+    return arr.reshape(*new_shape)
```

Comment:
TBH, I'm not sure this does the correct thing when the original shape is something like `(4,4,4,4,4)` and `axes` is something like `(1,3)`. Suggestions are more then welcome.

Reply:
> TBH, I'm not sure this does the correct thing when the original shape is something like `(4,4,4,4,4)` and `axes` is something like `(1,3)`. Suggestions are more then welcome.

I fixed that using `numpy.apply_along_axis` for every axis recursively.